### PR TITLE
[Move] Fully implement Tar Shot

### PR DESCRIPTION
**Pull request by torranx**
## What are the changes the user will see?
Tar Shot implemented

## Why am I making these changes?
partially implemented move

## What are the changes from a developer perspective?
new BattlerTag `DoubleMoveTypeEffectiveness`

## How to test the changes?
use Tar Shot then Fire-type move
or
`npm run test tar_shot`

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [ ] If I have text, did I make it translatable and add a key in the English locale file(s)?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [ ] Are the changes visual?
  - [ ] Have I provided screenshots/videos of the changes?

### CODE REVIEW

**Review by DayKev — DISMISSED** [review]

File: src/test/moves/tar_shot.test.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      .moveset([Moves.TAR_SHOT, Moves.FIRE_PUNCH])
+      .disableCrits();
+  });
+
+  it("lowers the target's Speed stat by one stage and doubles the effectiveness of Fire-type moves used on the target", async () => {
+    await game.classicMode.startBattle([Species.PIKACHU]);
+
+    const enemy = game.scene.getEnemyPokemon()!;
+
+    vi.spyOn(enemy, "getMaxHp").mockReturnValue(1000);
```
[the remark this answers is not in the document]
Comment (torranx) [reply src/test/moves/tar_shot.test.ts]: checked and it does not hehe

**Review by DayKev — APPROVED** [review]

**Review by Tempo-anon — DISMISSED** [review]

File: src/data/battler-tags.ts [hunk 2]
```diff
@@ -1984,6 +1984,30 @@ export class ExposedTag extends BattlerTag {
   }
 }
 
+/**
+ * Tag that doubles the type effectiveness of a move.
```
[the remark this answers is not in the document]
Comment (torranx) [reply src/data/battler-tags.ts:1988]: this is my lack of knowledge on moves - "there might be some move that is similar to how this works" so I named it like so. I can specify the name to `TarShotTag` instead if its the only move that does this.

[the remark this answers is not in the document]
Comment (Tempo-anon) [reply src/data/battler-tags.ts:1988]: You can keep it as is for now if you just note in the comments that it's a bespoke battlertag. This is more of a readability thing where bespoke battlertags with names that do not match up to the only instance they're used for is confusing for people unfamiliar with the code or have an encyclopedic knowledge of the game. We don't have a guideline for the exact naming convention of classes like this yet though my initial thought is that classes only used for one thing should reflect or document that

**Review by Tempo-anon — DISMISSED** [review]

**Comment by torranx** [comment]
Added message @frutescens , also fyi @Tempo-anon I changed the tag name to be specific (TarShotTag)

File: src/data/battler-tags.ts [hunk 3]
```diff
@@ -1984,6 +1984,29 @@ export class ExposedTag extends BattlerTag {
...
+   * If the Pokemon is terastallized, the tag cannot be added.
+   * @param pokemon the pokemon to which the tag is added
+   * @returns whether the tag is applied
+   */
+  canAdd(pokemon: Pokemon): boolean {
+    return !pokemon.isTerastallized();
+  }
+
+  onAdd(pokemon: Pokemon): void {
+    pokemon.scene.queueMessage(i18next.t("battlerTags:tarShotOnAdd", { pokemonNameWithAffix: getPokemonNameWithAffix(pokemon) }));
```
Comment (frutescens) [inline src/data/battler-tags.ts:2006]: This is more of a nit but I think 'wild' should be lowercase.

https://github.com/user-attachments/assets/183342f0-76a1-4876-b1fb-d911a3238543

**Review by frutescens — CHANGES_REQUESTED** [review]

File: src/data/battler-tags.ts (hunk 3 above)
Comment (frutescens) [reply src/data/battler-tags.ts:2006]: Or I don't know... it seems more of an issue with getPokemonNameWithAffix()
There really should be an option to return the affix as lowercase...

Comment (torranx) [reply src/data/battler-tags.ts:2006]: remove the starting word "The" instead to align with other text

**Review by frutescens — CHANGES_REQUESTED** [review]
Looks good to me.
Tested with some outlier scenarios.
This happened with Flash Fire but I'm not sure if Flash Fire has a message? 

https://github.com/user-attachments/assets/5f5d95d2-8e65-482b-b21b-1b00541bdf80

Otherwise, Clear Body, Contrary, and Well-Baked Body worked as expected.

**Comment by frutescens** [comment]
This does seem like an unrelated bug from Tar Shot so I'm going to report it to Issues instead.

**Review by frutescens — APPROVED** [review]

**Comment by torranx** [comment]
> This happened with Flash Fire but I'm not sure if Flash Fire has a message?

which one? seems ok to me?

**Comment by frutescens** [comment]
Ah nevermind.

**Comment by frutescens** [comment]
https://github.com/user-attachments/assets/2a693715-2bf2-4853-8829-36a3848a8537

This was being weird but it's unrelated to Tar Shot

File: src/field/pokemon.ts [hunk 4]
```diff
@@ -1372,6 +1372,10 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
       typeMultiplier.value = 0;
     }
 
+    if (this.getTag(TarShotTag) && (this.getMoveType(move) === Type.FIRE)) {
```
Comment (flx-sta) [inline src/field/pokemon.ts:1375]:
```suggestion
    if (this.hasTag(TarShotTag) && (this.getMoveType(move) === Type.FIRE)) {
```
I think in this case you would use `hasTag()`. Same result but clearer on whats' going on

File: src/test/moves/tar_shot.test.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
+import { BattlerIndex } from "#app/battle.js";
+import { Type } from "#app/data/type.js";
+import { Moves } from "#app/enums/moves.js";
+import { Species } from "#app/enums/species.js";
+import { Stat } from "#app/enums/stat.js";
```
Comment (flx-sta) [inline src/test/moves/tar_shot.test.ts:5]: The classic
```suggestion
import { BattlerIndex } from "#app/battle";
import { Type } from "#app/data/type";
import { Moves } from "#app/enums/moves";
import { Species } from "#app/enums/species";
import { Stat } from "#app/enums/stat";
```

File: src/data/battler-tags.ts [hunk 6]
```diff
@@ -1984,6 +1984,29 @@ export class ExposedTag extends BattlerTag {
...
+ * @extends BattlerTag
+ */
+export class TarShotTag extends BattlerTag {
+  constructor() {
+    super(BattlerTagType.TAR_SHOT, BattlerTagLapseType.CUSTOM, 0);
+  }
+
+  /**
+   * If the Pokemon is terastallized, the tag cannot be added.
+   * @param pokemon the pokemon to which the tag is added
```
Comment (flx-sta) [inline src/data/battler-tags.ts:1998]: NIT:
```suggestion
   * @param pokemon the {@linkcode Pokemon} to which the tag is added
```

File: src/data/battler-tags.ts [hunk 7]
```diff
@@ -1984,6 +1984,29 @@ export class ExposedTag extends BattlerTag {
...
+   * @returns whether the tag is applied
+   */
+  canAdd(pokemon: Pokemon): boolean {
```
Comment (flx-sta) [inline src/data/battler-tags.ts:2001]: NIT
```suggestion
  override canAdd(pokemon: Pokemon): boolean {
```

File: src/data/battler-tags.ts [hunk 8]
```diff
@@ -1984,6 +1984,29 @@ export class ExposedTag extends BattlerTag {
...
+    return !pokemon.isTerastallized();
+  }
+
+  onAdd(pokemon: Pokemon): void {
```
Comment (flx-sta) [inline src/data/battler-tags.ts:2005]: NIT:
```suggestion
  override onAdd(pokemon: Pokemon): void {
```

File: src/data/battler-tags.ts [hunk 9]
```diff
@@ -1984,6 +1984,29 @@ export class ExposedTag extends BattlerTag {
...
+    pokemon.scene.queueMessage(i18next.t("battlerTags:tarShotOnAdd", { pokemonNameWithAffix: getPokemonNameWithAffix(pokemon) }));
+  }
+}
+
 
 export function getBattlerTag(tagType: BattlerTagType, turnCount: number, sourceMove: Moves, sourceId: number): BattlerTag {
```
Comment (flx-sta) [inline src/data/battler-tags.ts:2011]: NIT:
Mind adding js-docs here?

**Comment by flx-sta** [comment]
Apart from my comments/NITs it looks to be working as expected:

### NOT Terastallized

<details>
  <summary>⏯️ NOT Terastallized Video</summary>

https://github.com/user-attachments/assets/59e164f7-d776-49c9-8955-09ea5cea3049

</details>


### Terastallized

<details>
  <summary>⏯️ Terastallized Video</summary>

https://github.com/user-attachments/assets/4c6a87d0-64dd-4486-a761-2ef236cd2f98

</details>

File: src/field/pokemon.ts (hunk 4 above)
Comment (torranx) [reply src/field/pokemon.ts:1375]: we do not have `hasTag` for `Pokemon` object

File: src/data/battler-tags.ts (hunk 7 above)
Comment (torranx) [reply src/data/battler-tags.ts:2001]: I will update the code with this but bear in mind we are not currently using the `override` keyword for other classes so an update to them would be somewhat `necessary` to be consistent

File: src/data/battler-tags.ts [hunk 10]
```diff
@@ -1984,7 +1984,38 @@ export class ExposedTag extends BattlerTag {
...
+    return !pokemon.isTerastallized();
+  }
+
+  override onAdd(pokemon: Pokemon): void {
+    pokemon.scene.queueMessage(i18next.t("battlerTags:tarShotOnAdd", { pokemonNameWithAffix: getPokemonNameWithAffix(pokemon) }));
+  }
+}
 
+/**
+ * Retrieves a BattlerTag based on the provided tag type, turn count, source move, and source ID.
```
Comment (flx-sta) [inline src/data/battler-tags.ts:2011]: NIT:
```suggestion
 * Retrieves a {@linkcode BattlerTag} based on the provided tag type, turn count, source move, and source ID.
```

**Review by flx-sta — APPROVED** [review]
lgtm (1 NIT)

**Review by flx-sta — APPROVED** [review]

**Review by frutescens — APPROVED** [review]

**Review by DayKev — APPROVED** [review]

**Review by CodeTappert — APPROVED** [review]
Approved only the new localization key

**Pull request merged**